### PR TITLE
[ZEPPELIN-960] When there is no interpreter, paragraph runJobapi  modified.

### PR DESCRIPTION
**Pull request by cloverhearts**
### What is this PR for?
Among Zeppelin API Fixed runNote.
According to whether the interpreter specified, 
was added to an exception processing for a result and operation.

run success (HTTP STATUS : 200)
```
{
  "status": "OK"
}
```

can't not found note id (HTTP STATUS : 404)
```
{
  "status": "NOT_FOUND",
  "message": "note not found."
}
```

interpter not found or not bind (HTTP STATUS : 412 -  PRECONDITION FAILED)
```
{
  "status": "PRECONDITION_FAILED",
  "message": "paragraph_1469771130099_-278315611 Not selected or Invalid Interpreter bind"
}
```

### What type of PR is it?
Bug Fix

### Todos
- [x] runNote api can apply by jdbc alias naming.
- [x] The results are recorded in the paragraph.
- [x] Binding fails when the interpreter 'HTTP STATUS CODE (412: PRECONDITION FAILED) and returns a message.
- [x] modification docs.

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-960

### How should this be tested?
execute to curl command line.
1. curl -XPOST http://(your zeppelin ip):(port)/api/notebook/job/2A94M5J1Z -H'Content-Type:application/json'
```
curl -XPOST http://127.0.0.1:8080/api/notebook/job/2A94M5J1Z -H'Content-Type:application/json'
```
2. 
if correct for execution then 
```
{"status":"OK"}
```
if invalid notebook id then
```
{
  "status": "NOT_FOUND",
  "message": "note not found."
}
```

if not binding interpreters
notebind

```
{
  "status": "PRECONDITION_FAILED",
  "message": "paragraph_(paragraphid) Not selected or Invalid Interpreter bind"
}
```

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? yes

### CODE REVIEW

**Comment by cloverhearts** [comment]
ping

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java [hunk 1]
```diff
@@ -498,7 +500,23 @@ public Response runNoteJobs(@PathParam("notebookId") String notebookId)
       return new JsonResponse<>(Status.NOT_FOUND, "note not found.").build();
     }
 
-    note.runAll();
+    for (Paragraph paragraph : note.getParagraphs()) {
+      note.setLastReplName(paragraph.getId());
```
Comment (jongyoul) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java]: I think it should be located after `note.run`

**Comment by jongyoul** [comment]
I suggest you'd better refactor `runAll` in order to call `run(paragraph)`. `runAll` actually submits all paragraphs into scheduler. How about it? In a api part, it looks good to me.

**Comment by cloverhearts** [comment]
ping

**Comment by cloverhearts** [comment]
retry ci

**Comment by cloverhearts** [comment]
@jongyoul 
You have completed the modifications as comments.
Please, check to this pr.
Thank you!

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java [hunk 2]
```diff
@@ -445,11 +445,11 @@ public void runAll() {
...
-
-        p.setListener(jobListenerFactory.getParagraphJobListener(this));
-        Interpreter intp = factory.getInterpreter(getId(), p.getRequiredReplName());
-
-        intp.getScheduler().submit(p);
+        try {
+          run(p.getId());
+        } catch (InterpreterException intpException) {
+          throw intpException;
+        }
```
Comment (bzz) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java]: What is the point of `try..catch` block here? 

Could you please explain, how is this code different from: 

```java
run(p.getId());
```

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java [hunk 3]
```diff
@@ -472,7 +472,14 @@ public void run(String paragraphId) {
         logger.debug("New paragraph: {}", pText);
         p.setEffectiveText(pText);
       } else {
-        throw new InterpreterException("Interpreter " + requiredReplName + " not found");
+        InterpreterException intpException;
+        intpException = new InterpreterException(
```
Comment (bzz) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java]: Probably a nitpick, but declaration and definition of this variable can be done on a single line.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java [hunk 4]
```diff
@@ -472,7 +472,14 @@ public void run(String paragraphId) {
...
+          "Pargaraph " + p.getJobName() + "'s Interpreter " + requiredReplName + " not found");
+        p.setReturn(
+          new InterpreterResult(InterpreterResult.Code.ERROR, intpException.getMessage()),
```
Comment (bzz) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java]: Again, probably a nitpick, but a variable extracted here, i.e  `intpResult` would most probably make code more readable.

```java
p.setReturn(intpResult);
p.setStatus(Job.Status.ERROR);
```

**Comment by bzz** [comment]
Great work @cloverhearts ! Few minor comments above.

Also, looks like CI is failing, and if you could post here the reason - other people could help determining if that is relevant to your change, or can be ignored.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java (hunk 2 above)
Comment (cloverhearts) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java]: @bzz 
Thank you for your feedback.
I wanted to express the `Exception` section occurs.
try .. catch means there is no action.
Does this part adjusts the confusion? If you have a problem, I would remove the `try..catch`.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java (hunk 3 above)
Comment (cloverhearts) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java]: @bzz 
It made by dividing the problem on the column size limit 100.
I have to fix it again.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java (hunk 4 above)
Comment (cloverhearts) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java]: Modification was completed.
Thank you.

**Comment by cloverhearts** [comment]
Thank you @bzz !
CI Test of failure seems to be unrelated to my code.
It will probably think of other problems.

**Comment by cloverhearts** [comment]
ping

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java (hunk 2 above)
Comment (bzz) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Note.java]: Yes, basically, we either better either remove `try..catch` completely or refactor it to catch more specific exception + document why there is nothing happening in catch. See [Zeppelin's Java styleguide](https://google.github.io/styleguide/javaguide.html#s6.2-caught-exceptions) for more detailed explanation.

**Comment by bzz** [comment]
Thank you for your patience!
Looks good to me, except minor style issue above?

As soon as it is addressed - I think we should be good to merge it to master.

**Comment by cloverhearts** [comment]
@bzz Thank you.
I had remove to try catch.
All requests have been completed.
Have a Nice Day--!

**Comment by bzz** [comment]
Thank you @cloverhearts !
Looks good to me, merging if there is no further discussion

**Pull request closed without merging**